server/src/main/java/org/elasticsearch/index/engine/LuceneChangesSnapshot.java

```diff
@@ -0,0 +1,213 @@
+import org.elasticsearch.index.translog.Translog;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.List;
+import java.util.function.Supplier;
+
+/**
+ * A {@link Translog.Snapshot} from changes in a Lucene index
+ */
+final class LuceneChangesSnapshot implements Translog.Snapshot {
+    private final long fromSeqNo, toSeqNo;
+    private long lastSeenSeqNo;
+    private int skippedOperations;
+    private final boolean requiredFullRange;
+
+    private final IndexSearcher searcher;
+    private final MapperService mapperService;
+    private int docIndex;
+    private final TopDocs topDocs;
+
+    private final Closeable onClose;
+
+    /**
+     * Creates a new "translog" snapshot from Lucene for reading operations whose seq# in the specified range.
+     *
+     * @param searcherFactory   the engine searcher factory (prefer the internal searcher)
+     * @param mapperService     the mapper service which will be mainly used to resolve the document's type and uid
+     * @param fromSeqNo         the min requesting seq# - inclusive
+     * @param toSeqNo           the maximum requesting seq# - inclusive
+     * @param requiredFullRange if true, the snapshot will strictly check for the existence of operations between fromSeqNo and toSeqNo
+     */
+    LuceneChangesSnapshot(Supplier<Engine.Searcher> searcherFactory, MapperService mapperService,
+                          long fromSeqNo, long toSeqNo, boolean requiredFullRange) throws IOException {
+        if (fromSeqNo < 0 || toSeqNo < 0 || fromSeqNo > toSeqNo) {
+            throw new IllegalArgumentException("Invalid range; from_seqno [" + fromSeqNo + "], to_seqno [" + toSeqNo + "]");
+        }
+        this.mapperService = mapperService;
+        this.fromSeqNo = fromSeqNo;
+        this.toSeqNo = toSeqNo;
+        this.lastSeenSeqNo = fromSeqNo - 1;
+        this.requiredFullRange = requiredFullRange;
+        boolean success = false;
+        final Engine.Searcher engineSearcher = searcherFactory.get();
+        try {
+            this.searcher = new IndexSearcher(Lucene.wrapAllDocsLive(engineSearcher.getDirectoryReader()));
+            this.searcher.setQueryCache(null);
+            this.topDocs = searchOperations(searcher);
+            success = true;
+            this.onClose = engineSearcher;
+        } finally {
+            if (success == false) {
+                IOUtils.close(engineSearcher);
+            }
+        }
+    }
+
+    @Override
+    public void close() throws IOException {
+        onClose.close();
+    }
+
+    @Override
+    public int totalOperations() {
+        return Math.toIntExact(topDocs.totalHits);
+    }
+
+    @Override
+    public int overriddenOperations() {
+        return skippedOperations;
+    }
+
+    @Override
+    public Translog.Operation next() throws IOException {
+        final Translog.Operation op = nextOp();
+        if (requiredFullRange && lastSeenSeqNo < toSeqNo) {
```

Comment:
why we to check here that lastSeenSeqNo is < toSeqNo? shouldn't we stop reading before this happens?

Reply:
Do we also want to assert that seqNo != lastSeeSeqNo?

Reply:
The caller should continue consuming the snapshot until the `next` method returns null. In the last call, lastSeenSeqNo equals to toSeqNo and op is null. This guard is added to avoid checking in this case. I am +1 on the assertion.